clients/src/main/java/org/apache/kafka/common/config/AbstractConfig.java

```diff
@@ -369,6 +401,68 @@ public void logUnused() {
         return objects;
     }
 
+    /**
+     * Instantiates given list of config providers and fetches the actual values of config variables from the config providers.
+     * returns a map of config key and resolved values.
+     * @param indirectVariables The map of config variables
+     * @param configProviders The map of config provider configs
+     * @param original The map of raw configs.
+     * @return map of resolved config variable.
+     */
+    @SuppressWarnings("unchecked")
+    private  Map<String, ?> resolveConfigVariables(Map<String, String> indirectVariables, Map<?, ?> configProviders, Map<String, Object> original) {
+        Map<String, ConfigProvider> providers;
+
+        if (configProviders == null || configProviders.isEmpty()) {
+            providers = instantiateConfigProvider(indirectVariables);
+        } else {
+
+            for (Map.Entry<?, ?> entry : configProviders.entrySet()) {
+                if (!(entry.getKey() instanceof String))
+                    configProviders.remove(entry.getKey());
+            }
+            providers = instantiateConfigProvider((Map<String, ?>) configProviders);
+        }
+        if (!providers.isEmpty()) {
+            ConfigTransformer configTransformer = new ConfigTransformer(providers);
+            ConfigTransformerResult result = configTransformer.transform(indirectVariables);
+            original.putAll(result.data());
+        }
+
+        return original;
+    }
+
+    private Map<String, ConfigProvider> instantiateConfigProvider(Map<String, ?> indirectConfigs) {
+        final String configProviders = (String) indirectConfigs.get(CONFIG_PROVIDERS_CONFIG);
+
+        if (configProviders == null || configProviders.isEmpty())
+            return Collections.emptyMap();
+
+        Map<String, String> providerMap = new HashMap<>();
+
+        for (String provider: configProviders.split(",")) {
+            String providerClass = CONFIG_PROVIDERS_CONFIG + "." + provider + ".class";
+            if (indirectConfigs.containsKey(providerClass))
+                providerMap.put(provider, (String) indirectConfigs.get(providerClass));
+
+        }
+        // Instantiate Config Providers
+        Map<String, ConfigProvider> configProviderInstances = new HashMap<String, ConfigProvider>();
+        for (Map.Entry<?, ?> entry : providerMap.entrySet()) {
+            if ((entry.getKey() instanceof String) && (entry.getValue() instanceof String)) {
```

Comment:
providerMap is Map<String, String> , so these will be Strings, see comment above.